File: day-8/main.py
inputGrid = []

with open('input', 'r') as file:
    for line in file:
        inputGrid.append([int(c) for c in line.strip()])

## Iterate once through to find all trees visible from top or left
treesVisible = []
tallestTreeVertical = [-1 for i in inputGrid[0]]
for idy, row in enumerate(inputGrid):
    tallestTreeHorizontal = -1
    treesVisible.append([])
    for idx, tree in enumerate(row):
        visibility = 0

        # Tree is visible from left
        if tree > tallestTreeHorizontal:
            tallestTreeHorizontal = tree
            visibility = 1

        # Tree is visible from top
        if tree > tallestTreeVertical[idx]:
            tallestTreeVertical[idx] = tree
            visibility = 1

        treesVisible[idy].append(visibility)

# Iterate through again (reversed) to find trees visible from right or bottom
tallestTreeVertical = [-1 for i in inputGrid[0]]
for idy, row in reversed(list(enumerate(inputGrid))):
    tallestTreeHorizontal = -1
    for idx, tree in reversed(list(enumerate(row))):
        visibility = 0

        # Tree is visible from left
        if tree > tallestTreeHorizontal:
            tallestTreeHorizontal = tree
            visibility = 1

        # Tree is visible from top
        if tree > tallestTreeVertical[idx]:
            tallestTreeVertical[idx] = tree
            visibility = 1

        treesVisible[idy][idx] |= visibility

# count all 1s for part 1
print('Part 1: {}'.format(sum([sum(row) for row in treesVisible])))

File: day-8/main.ts
let inputGrid: number[][] = [];

fetch('input')
  .then(response => response.text())
  .then(data => {
    const lines = data.trim().split('\n');
    inputGrid = lines.map(line => line.trim().split('').map(Number));
    calculateTreesVisible();
  });

function calculateTreesVisible() {
  const treesVisible: number[][] = [];
  const tallestTreeVertical: number[] = Array(inputGrid[0].length).fill(-1);

  for (let idy = 0; idy < inputGrid.length; idy++) {
    let tallestTreeHorizontal = -1;
    treesVisible.push([]);

    for (let idx = 0; idx < inputGrid[idy].length; idx++) {
      let visibility = 0;

      // Tree is visible from left
      if (inputGrid[idy][idx] > tallestTreeHorizontal) {
        tallestTreeHorizontal = inputGrid[idy][idx];
        visibility = 1;
      }

      // Tree is visible from top
      if (inputGrid[idy][idx] > tallestTreeVertical[idx]) {
        tallestTreeVertical[idx] = inputGrid[idy][idx];
        visibility = 1;
      }

      treesVisible[idy].push(visibility);
    }
  }

  tallestTreeVertical.fill(-1);

  for (let idy = inputGrid.length - 1; idy >= 0; idy--) {
    let tallestTreeHorizontal = -1;

    for (let idx = inputGrid[idy].length - 1; idx >= 0; idx--) {
      let visibility = 0;

      // Tree is visible from left
      if (inputGrid[idy][idx] > tallestTreeHorizontal) {
        tallestTreeHorizontal = inputGrid[idy][idx];
        visibility = 1;
      }

      // Tree is visible from top
      if (inputGrid[idy][idx] > tallestTreeVertical[idx]) {
        tallestTreeVertical[idx] = inputGrid[idy][idx];
        visibility = 1;
      }

      treesVisible[idy][idx] |= visibility;
    }
  }

  const sumVisibleTrees = treesVisible.reduce((acc, row) => acc + row.reduce((a, b) => a + b, 0), 0);
  console.log(`Part 1: ${sumVisibleTrees}`);
}
